fasm/output.py

```diff
@@ -0,0 +1,312 @@
+import enum
+from fasm import SetFasmFeature, FasmLine, ValueFormat
+
+def is_only_comment(line):
+    """ Returns True if line is only a comment. """
+    return not line.set_feature and not line.annotations and line.comment
+
+def is_only_annotation(line):
+    """ Returns True if line is only an annotations. """
+    return not line.set_feature and line.annotations and not line.comment
+
+def is_blank_line(line):
+    """ Returns True if line is blank. """
+    return not line.set_feature and not line.annotations and not line.comment
+
+def merge_features(features):
+    """ Combines features with varying addresses but same feature into one feature.
+
+    A[0] = 1
+    A[1] = 1
+
+    becomes
+
+    A[1:0] = 2'b11
+
+    and
+
+    A[5] = 1
+    A[7] = 1
+
+    A[7:0] = 8'b10100000
+
+    """
+    # Ensure all features are for the same feature
+    assert len(set(feature.feature for feature in features)) == 1
+
+    set_bits = set()
+    cleared_bits = set()
+
+    for feature in features:
+        start = 0
+        end = 0
+        value = 1
+
+        if feature.start is None:
+            assert feature.end is None
+        else:
+            start = feature.start
+            if feature.end is not None:
+                end = feature.end
+            else:
+                end = start
+
+        if feature.value is not None:
+            value = feature.value
+
+        for bit in range(start, end+1):
+            bit_is_set = ((value >> (bit - start)) & 1) != 0
+            if bit_is_set:
+                assert bit not in cleared_bits
+                set_bits.add(bit)
+            else:
+                assert bit not in set_bits
+                cleared_bits.add(bit)
+
+    max_bit = max(set_bits, cleared_bits)
+
+    final_value = 0
+    for bit in set_bits:
+        final_value |= (1 << bit)
+
+    return SetFasmFeature(
+            feature=features[0].feature,
+            start=0,
+            end=max_bit,
+            value=final_value,
+            value_format=ValueFormat.VERILOG_BINARY)
+
+class MergeModel(object):
+    """ Groups and merges features.
+
+    Grouping logic:
+     - Consecutive comments will be grouped.
+     - Comments groups will attach to the next non-comment entry.
+     - Consecutive annotations will be grouped.
+     - Empty lines will be discarded
+     - Features will be grouped by their first feature part.
+     - Features within the same feature with different addresses will be merged.
+
+    If a feature has a comment in its group, it is not eligable for address merging.
+
+    """
+    class State(enum.Enum):
+        NoGroup = 1
+        InCommentGroup = 2
+        InAnnotationGroup = 3
+
+    def __init__(self):
+        self.state = MergeModel.State.NoGroup
+        self.groups = []
+        self.current_group = None
+
+    def start_comment_group(self, line):
```

Comment:
(optional) These functions could use doc strings.

Reply:
Added comments for the public ones.